qiskit/quantum_info/synthesis/two_qubit_decompose.py

```diff
@@ -926,9 +949,288 @@ def __call__(self, target, basis_fidelity=None, *, _num_basis_uses=None) -> Quan
+            euler_q0[iqubit, [1, 2, 0]] = euler_angles[:3]
+            global_phase += euler_angles[3]
+        # decompose target unitaries to xzx
+        xzx_decomposer = OneQubitEulerDecomposer("XZX")
+        for iqubit, decomp in enumerate(decomposition[1::2]):
+            euler_angles = xzx_decomposer.angles_and_phase(decomp)
+            euler_q1[iqubit, [1, 2, 0]] = euler_angles[:3]
+            global_phase += euler_angles[3]
+
+        qc = QuantumCircuit(2)
+        qc.global_phase = target_decomposed.global_phase
+        qc.global_phase -= best_nbasis * self.basis.global_phase
+        qc.global_phase += global_phase
+
+        x12 = euler_q0[1][2] + euler_q0[2][0]
+        x12_isNonZero = not math.isclose(x12, 0, abs_tol=atol)
+        x12_isPi = math.isclose(x12, math.pi, abs_tol=atol)
+        x10_isPi = math.isclose(euler_q0[1][0], math.pi, abs_tol=atol)
+        x12_isNegPi = math.isclose(x12, -math.pi, abs_tol=atol)
+        x10_isZero = math.isclose(euler_q0[1][0], 0, abs_tol=atol)
+        x12_isHalfPi = math.isclose(x12, math.pi / 2, abs_tol=atol)
+        # TODO: make this more effecient to avoid double decomposition
+        circ = QuantumCircuit(1)
+        circ.rz(euler_q0[0][0], 0)
+        circ.rx(euler_q0[0][1], 0)
+        if x12_isPi and x10_isPi:
+            circ.rz(euler_q0[0][2], 0)
+            circ.global_phase += math.pi
+        elif x12_isNegPi and x10_isZero:
+            # -pi donated to next rz, which vanishes in conjugation of rx
+            circ.rz(euler_q0[0][2] + math.pi, 0)
+            circ.global_phase += math.pi
+        else:
+            circ.rz(euler_q0[0][2] + euler_q0[1][0], 0)
+        circ.h(0)
+        qceuler = self._decomposer1q(Operator(circ).data)
+        qc.compose(qceuler, [0], inplace=True)
+
+        circ = QuantumCircuit(1)
+        circ.rx(euler_q1[0][0], 0)
+        circ.rz(euler_q1[0][1], 0)
+        circ.rx(euler_q1[0][2] + euler_q1[1][0], 0)
+        circ.h(0)
+        qceuler = self._decomposer1q(Operator(circ).data)
+        qc.compose(qceuler, [1], inplace=True)
+
+        qc.cx(1, 0)
+
+        if (x12_isPi and x10_isPi) or (x12_isNegPi and x10_isZero):
+            qc.rz(-euler_q0[1][1], 0)
+        else:
+            qc.rz(euler_q0[1][1], 0)
+        if x12_isHalfPi:
+            qc.sx(0)
+            qc.global_phase -= math.pi / 4
+        elif x12_isNonZero and not ((x12_isPi and x10_isPi) or (x12_isNegPi and x10_isZero)):
+            # this is non-optimal but doesn't seem to occur currently
+            if self.pulse_optimize is None:
+                qc.compose(self._decomposer1q(Operator(RXGate(x12)).data), [0], inplace=True)
+            else:
+                print_logic_flags()
+                print(euler_q0)
+                print(euler_q1)
+                print(get_basic_decomp(euler_q0, euler_q1))
+                raise QiskitError("possible non-pulse-optimal decomposition encountered")
```

Comment:
It looks like a handful of cases of `TestPulseOptimalDecompose.test_sx_virtz_3cnot_optimal`,  `TestUnitarySynthesis.test_qv_natural` and a few others are failing here: https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=29862&view=logs&j=7a68dc58-ebf8-506c-1eea-de91a393131c&t=cbde9f75-c240-5751-2949-a776e1c24901&l=14618

Reply:
i think this check is unnecessary. 
a) I checked test_sx_virtz_3cnot_optimal and it is actually doing the pulse optimized decomposition in all cases, so this check is not correct.
b) Even if the decomposer does not find the optimal solution in some case, I don't think it should raise. It is not clear whether the solutions we have are "optimal" (although if your metric is number of SX pulses it probably is). There could be other pulse decompositions that are even better. Raising an error is not good behavior for an optimizer I think, all of this stuff is best effort.

Reply:
That test is supposed to catch if something was unexpected in the decomp and pulse_optimize was set to True. Since None is the default it should never raise for legacy code. One would have to explicitly be setting the option to True in a new pass manager. Also, it seems to be not raising on all platforms in CI currently. I set it True in the test but could probably relax it to None there.